cryptol-style.md

```diff
@@ -0,0 +1,405 @@
+types on separate lines.
+
+```haskell
+sayHello :
+     {a}
+     (fin a) =>
+     [a][8] -> [a+7][8]
+```
+
+The `:` should always be surrounded by a single space on either
+side. This aligns type definitions with value definitions (where the `=`
+is also surrounded by a single space on either side). For example,
+
+```haskell
+x : [32]
+x = 10
+y : [4][32]
+y = [1, 2, 3, 4]
+```
+
+### Type Constraints
+
+Type constraints should always be tupleized. The following is incorrect:
+
+```haskell
+fun :
+   {a, b}
+   fin a => fin b =>
+   [a][b] -> [a+b]
+```
+
+`fun` should be:
+
+```haskell
+fun :
+   {a, b}
+   (fin a, fin b)
+   [a][b] -> [a+b]
+```
+
+### Data Declarations
+
+The `=` should always be surrounded by a single space on either
+side. This aligns value definitions with type definitions (where the `:`
+is also surrounded by a single space on either side). For example,
+
+```haskell
+x : [32]
+x = 10
+y : [4][32]
+y = [1, 2, 3, 4]
+```
+
+Format records as follows:
+
+```haskell
+Person =
+    { firstName : String 10
+    , lastName  : String 10
+    , age       : Integer
+    }
+```
+
+Align the elements in a list. For example,
+
+```haskell
+exceptions =
+    [ InvalidStatusCode
+    , MissingContentHeader
+    , InternalServerError
+    ]
+```
+
+Optionally, you can skip the first newline. Use your judgement or try
+to match the paper document of the specification you're working from.
+
+```haskell
+directions = [ North
+             , East
+             , South
+             , West
+             ]
+```
+
+
+### If-then-else clauses
+
+Align if-then-else clauses like you would normal expressions:
+
+```haskell
+foo = if ...
+      then ...
+      else ...
+```
+
+The same rule applies in nested where clauses:
+
+```haskell
+foo a b = c
+  where
+    c = if a == 0x0000
+        then d
+          where
+            d = if b == 0x0000
+                then 0x1000
+                else 0x0010
+        else 0x0100
+```
+
+### Where clauses
+
+Align the `=` symbols in a where clause when it helps readability.
+
+```haskell
+foo = ...
+  where
+    cat   = ...
+    fish  = ...
+    dog   = ...
+    horse = ...
+```
+
+Comments
+--------
+
+### Punctuation
+
+Write proper sentences; start with a capital letter and use proper
+punctuation.
+
+### Top-Level Definitions
+
+Comment every top level function (particularly exported functions),
+and provide a type signature. The documentation should
+give enough information to
+apply the function without looking at the function's definition.
+
+When writing a cryptoalgorithm specification in Cryptol, often the
+arguments to a function have verbiage in the specification that may be
+copied into the function documentation.
+This is considered a good practice provided it does not become too verbose.
+
+### End-of-Line Comments
+
+Separate end-of-line comments from the code using 2 spaces.
+
+```haskell
+foo : [32] -> [32]
+foo n = n % p
+  where
+    p = 4294967291  -- Largest 32 bit prime.
+```
```

Comment:
End of line comments aren't referenced in the Programming Cryptol document (https://cryptol.net/files/ProgrammingCryptol.pdf), and my lame attempt with them was unsuccessful...